stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/avatar/AvatarBitmapFactory.kt

```diff
@@ -0,0 +1,283 @@
+    ): List<Bitmap> {
+        return users.take(MAX_AVATAR_SECTIONS).map { createUserBitmapInternal(it, style, avatarSize) }
+    }
+
+    private fun createInitialsBitmap(
+        avatarStyle: AvatarStyle,
+        @Px avatarSize: Int,
+        initials: String,
+    ): Bitmap {
+        val avatarSizeWithoutBorder = avatarSize - avatarStyle.avatarBorderWidth * 2
+        return Bitmap.createBitmap(
+            avatarSizeWithoutBorder,
+            avatarSizeWithoutBorder,
+            Bitmap.Config.ARGB_8888
+        ).apply {
+            val canvas = Canvas(this)
+            canvas.drawGradient(initials, avatarSizeWithoutBorder)
+            canvas.drawInitials(avatarStyle, initials, avatarSizeWithoutBorder)
+        }
+    }
+
+    private fun Canvas.drawGradient(initials: String, avatarSize: Int) {
+        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
+            isAntiAlias = true
+            style = Paint.Style.FILL
+            shader = createLinearGradientShader(initials, avatarSize)
+        }
+        drawRect(
+            0f,
+            0f,
+            avatarSize.toFloat(),
+            avatarSize.toFloat(),
+            paint
+        )
+    }
+
+    private fun Canvas.drawInitials(
+        avatarStyle: AvatarStyle,
+        initials: String,
+        @Px avatarSize: Int,
+    ) {
+        val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
+            style = Paint.Style.FILL
+            typeface = avatarStyle.avatarInitialText.font ?: Typeface.DEFAULT_BOLD
+            textAlign = Paint.Align.CENTER
+            color = avatarStyle.avatarInitialText.color
+            textSize = avatarStyle.avatarInitialText.size.toFloat()
+        }
+        drawText(
+            initials,
+            avatarSize / 2f,
+            avatarSize / 2f - (textPaint.ascent() + textPaint.descent()) / 2f,
+            textPaint
+        )
+    }
+
+    private fun createLinearGradientShader(initials: String, @Px avatarSize: Int): Shader {
+        val baseColorIndex = abs(initials.hashCode()) % gradientBaseColors.size
+        val baseColor = gradientBaseColors[baseColorIndex]
+        return LinearGradient(
+            0f,
+            0f,
+            0f,
+            avatarSize.toFloat(),
+            adjustColorLBrightness(baseColor, GRADIENT_DARKER_COLOR_FACTOR),
+            adjustColorLBrightness(baseColor, GRADIENT_LIGHTER_COLOR_FACTOR),
+            Shader.TileMode.CLAMP
+        )
+    }
+
+    public companion object {
+        private const val GRADIENT_DARKER_COLOR_FACTOR = 1.3f
+        private const val GRADIENT_LIGHTER_COLOR_FACTOR = 0.7f
+
+        /**
+         * Marker object to detect whether methods have been implemented by custom subclasses.
+         */
+        private val NOT_IMPLEMENTED_MARKER = Bitmap.createBitmap(0, 0, Bitmap.Config.ALPHA_8)
+        internal lateinit var instance: AvatarBitmapFactory
+
+        public fun overrideDefaultInstance(avatarBitmapFactory: AvatarBitmapFactory) {
```

Comment:
`setInstance` perhaps?

Reply:
It will have conflict with the property itself, should we remove this method and open the `instance` property on that case?

Reply:
BTW, we should move it to `ChatUI` when the refactor is completed

Reply:
`private set` for the property?